synapse_spamcheck_badlist/bad_list_filter.py

```diff
@@ -127,17 +130,19 @@ async def check_event_for_spam(self, event) -> bool:
             # Check for links in text, both unformatted and formatted.
             #
             # We always lower-case the url, as the IWF database is lowercase.
-            for text in [content.get("body", ""), content.get("formatted_body", "")]:
-                # Run a first, faster test.
-                if not self._linkifier.test(text):
-                    continue
-                # Now run the slower test, if necessary, using results cached from the faster test.
-                for match in self._linkifier.match(text) or []:
-                    link = re.sub(self._scheme_re, "", match.url.lower())
-                    is_bad_link = await self._api.run_db_interaction("Check link against evil db", _db_is_bad_link, self._links_table, link)
-                    if is_bad_link:
-                        logger.info("Rejected bad link")
-                        return True
+            with self._link_test_performance.time():
+                for text in [content.get("body", ""), content.get("formatted_body", "")]:
+                    # Run a first, faster test.
+                    with self._linkify_test_performance.time():
+                        if not self._linkifier.test(text):
+                            continue
+                        # Now run the slower test, if necessary, using results cached from the faster test.
```

Comment:
Is `test` that much faster than `match`? Especially if it is caching some result then, I don't see how having them be separate makes much sense.

Reply:
This is what the documentation of `linkify` recommends. Looking at the source code of `linkify`, I'm not entirely convinced, but I believe I should try and do what the doc says until I have a better idea of the performance profile.

At this stage, I see the following possible outcomes:
1. Performance is good, no need to investigate further;
2. Performance is a bit bad, let's see what we can do to improve it (e.g. what you're suggesting);
3. Performance is really, really bad, let's see if we can do better without `linkify`.